Gas Price Calculation

kdonthi opened the pull request.

LukaszRozmej commented on src/Nethermind/Nethermind.Core.Test/Builders/BlockBuilder.cs:
```diff
@@ -78,6 +78,11 @@ public BlockBuilder WithTransactions(int txCount, IReleaseSpec releaseSpec)
             return WithTransactions(txs);
         }
+
+        public BlockBuilder WithKnownTransactions(Transaction[] transactions)
+        {
+            return WithTransactions(transactions);
+        }
```
What is the purpose there?

kdonthi replied:
I wanted to be able to build a block with transactions I pass in. The alternative looked like "WithTransactions" where I can create a certain number of transactions, but I don't get to specify gas prices and block numbers.

Maybe I should rename it to "WithGivenTransactions"?

kdonthi replied:
Actually I just realized that this is pointless haha, I could have just used WithTransactions to begin with

LukaszRozmej commented on src/Nethermind/Nethermind.JsonRpc/Modules/Eth/EthRpcModule.cs:
```diff
@@ -140,12 +142,108 @@ public ResultWrapper<Address> eth_coinbase()
 
-        [Todo("Gas pricer to be implemented")]
-        public ResultWrapper<UInt256?> eth_gasPrice()
+        private bool add_transactions_from_block_to_set(Block block, ref SortedSet<UInt256> sortedSet, UInt256 finalPrice, UInt256? ignoreUnder = null, long? maxCount = null)
```
Method naming not within convention

kdonthi replied:
Are you referring to eth_gasPrice or add_transactions_from_block_to_set?

LukaszRozmej commented on src/Nethermind/Nethermind.JsonRpc/Modules/Eth/EthRpcModule.cs:
```diff
@@ -140,12 +142,108 @@ public ResultWrapper<Address> eth_coinbase()
+        [Todo("Gas pricer to be implemented")]
+        public ResultWrapper<UInt256?> eth_gasPrice(UInt256? ignoreUnder = null)
+        {
+            Block headBlock = _blockFinder.FindHeadBlock();
```
Move to some seperate class that will calculate gas price and keep last calculation so don't recalculate it if recalculation is not needed

kdonthi replied:
Is it fine if I have the logic for eth_gasPrice in a different class but be called from within EthRpcModule? 

I am using the _blockFinder to search for blocks when I do need to calculate the gasPrice and I would want to pass that in as a parameter.

LukaszRozmej reviewed (changes requested):
Some basic changes, lets have a review call after their fixed

LukaszRozmej commented on src/Nethermind/Nethermind.JsonRpc/Modules/Eth/EthRpcModule.cs:
```diff
@@ -140,12 +142,108 @@ public ResultWrapper<Address> eth_coinbase()
+            latestGasPrice = 1; //do we want to throw an error if there are no transactions? Do we want to set lastPrice to 1 when we cannot find latestPrice in tx?
         }
+        
+        [Todo("Gas pricer to be implemented")]
```
Remove Todo

LukaszRozmej commented on src/Nethermind/Nethermind.JsonRpc/Modules/Eth/EthRpcModule.cs:
```diff
@@ -140,12 +142,108 @@ public ResultWrapper<Address> eth_coinbase()
+            Comparer<UInt256> comparer = Comparer<UInt256>.Create(((a, b) =>
+            {
+                int res = a.CompareTo(b);
+                return res == 0 ? 1 : res;
+            })); 
+            SortedSet<UInt256> gasPrices = new(comparer); //allows duplicates
+            long blocksToGoBack = 5;
+            const int percentile = 20;
+            long threshold = blocksToGoBack * 2;
+            UInt256? gasPriceLatest = null;
+            latest_gas_price(headBlock!.Number, genesisBlock!.Number, ref gasPriceLatest);
+            if (gasPriceLatest == null)
+            {
+                return ResultWrapper<UInt256?>.Fail("gasPriceLatest was not set properly.");
+            }
+            long blockNumber = headBlock!.Number; 
 
+            while (blocksToGoBack > 0 && blockNumber > genesisBlock!.Number - 1) //else, go back "blockNumber" valid gas prices from genesisBlock
+            {
+                Block? foundBlock = _blockFinder.FindBlock(blockNumber);
+                if (foundBlock != null)
+                {
+                    bool result = add_transactions_from_block_to_set(foundBlock, ref gasPrices, (UInt256) gasPriceLatest, ignoreUnder);
+                    if (result || gasPrices.Count + blocksToGoBack >= threshold) //if we only added one transaction, we don't reduce blocksToGoBack if second condition holds
+                    {
+                        blocksToGoBack--;
+                    }
+                }
+                blockNumber--;
+            }
+
+            int finalIndex = (int) Math.Round(((gasPrices.Count - 1) * ((float) percentile / 100)));
+            foreach (UInt256 gasPrice in gasPrices.Where(_ => finalIndex-- <= 0))
+            {
+                gasPriceLatest = gasPrice;
+                break;
```
The logic is quite hard to follow, can this be simplified by better naming or more submethods?

marcindsobczak commented:
I tested it manually on Goerli, sent eth_gasPrice requests to Geth and Neth parallel, tried it several times at 2-3min intervals and results were always the same. Seems that your implementation works well, same as Geth.

MarekM25 commented on src/Nethermind/Nethermind.JsonRpc/Modules/Eth/EthRpcModule.cs:
```diff
@@ -140,12 +142,108 @@ public ResultWrapper<Address> eth_coinbase()
+        {
+            while (headBlockNumber >= genesisBlockNumber) //should this be latestBlock or headBlock?
+            {
+                Transaction[] transactions = _blockFinder.FindBlock(headBlockNumber)!.Transactions;
```
My understanding is that we should filter out EIP1559 transactions because their gas price field would be invalid. Please add a test case for that scenario.

MarekM25 replied:
Please check Peter comment: https://github.com/ethereum/pm/issues/328#issuecomment-853234014

tkstanczak commented on src/Nethermind/Nethermind.JsonRpc/Modules/Eth/GasPrice.cs:
```diff
@@ -0,0 +1,176 @@
+
+namespace Nethermind.JsonRpc.Modules.Eth
+{
+    public class GasPrice
```
name of this class does not represent what the class does.
Is this class a gas price? not
Does this class calculate the gas price to be returned by JSON RPC? yes

name should be more like MinGasPriceEstimator, GasPriceEstimator, GasPriceOracle

tkstanczak replied:
Ideally such a class would be behind an interface IGasPriceOracle
(gas price oracle is some magical provider of the current gas price)
Implementation can be internal, or based on GasNow.org etc

tkstanczak commented on src/Nethermind/Nethermind.JsonRpc/Modules/Eth/GasPrice.cs:
```diff
@@ -0,0 +1,176 @@
+{
+    public class GasPrice
+    {
+        private static UInt256? _lastPrice = null; //will this be okay if it is static?
```
this is not ok to be static, this should be an internal class field

tkstanczak commented on src/Nethermind/Nethermind.JsonRpc/Modules/Eth/GasPrice.cs:
```diff
@@ -0,0 +1,176 @@
+    public class GasPrice
+    {
+        private static UInt256? _lastPrice = null; //will this be okay if it is static?
+        private static Block? _lastHeadBlock = null;
```
same here

tkstanczak commented on src/Nethermind/Nethermind.JsonRpc/Modules/Eth/GasPrice.cs:
```diff
@@ -0,0 +1,176 @@
+        private readonly IBlockFinder _blockFinder;
+        public GasPrice(IBlockFinder blockFinder)
+        {
+            _blockFinder = blockFinder;
```
verify nullability as in other places in code, read about fail fast
I suggest to read the Clean Code book (C# style)

tkstanczak commented on src/Nethermind/Nethermind.JsonRpc/Modules/Eth/GasPrice.cs:
```diff
@@ -0,0 +1,176 @@
+        {
+            _blockFinder = blockFinder;
+        }
+        private bool eth_addTxFromBlockToSet(Block block, ref SortedSet<UInt256> sortedSet, UInt256 finalPrice,
```
If you look at the module, you will notice convention where eth_[something] names are only reserved for the methods related to the public JSON RPC endpoints. All other methods always follow standard naming from C#.
Try to look at the components that you write in and understand their conventions and logic.

tkstanczak commented on src/Nethermind/Nethermind.JsonRpc/Modules/Eth/GasPrice.cs:
```diff
@@ -0,0 +1,176 @@
+            return gasPrices;
+        }
+        
+        private static bool eth_initalChecks(Block? headBlock, Block? genesisBlock, out ResultWrapper<UInt256?> resultWrapper)
```
Your error message is inconsistent with your condition testing.
Did you want to write:
headBlock == null && genesisBlock == null


Would be much nicer to split it into two methods with very clear responsibility:

Handle

tkstanczak replied:
Would be much nicer to split it into two methods with very clear responsibility:

HandleMissingHeadBlockCase(...)
HandleNoHeadBlockChange(...)

tkstanczak commented on src/Nethermind/Nethermind.JsonRpc/Modules/Eth/GasPrice.cs:
```diff
@@ -0,0 +1,176 @@
+
+        private UInt256? eth_latestGasPrice(long headBlockNumber, long genesisBlockNumber)
+        {
+            while (headBlockNumber >= genesisBlockNumber) //should this be latestBlock or headBlock?
```
genesis block always has number 0

also a loop like this -> you need to limit this loop since now it will go very far back in time on chains that have many empty blocks -> ideally you would limit the search to 8 / 16 blocks and in such cases if there have been no transactions then you should set the price to minimum

tkstanczak commented on src/Nethermind/Nethermind.JsonRpc/Modules/Eth/GasPrice.cs:
```diff
@@ -0,0 +1,176 @@
+                    .Where(t => !t.IsEip1559).ToArray();
+                if (transactions.Length > 0)
+                {
+                    return transactions[^1].GasPrice;
```
this is minimum gas price from previous non-empty block, did not you want to go slightly above this?

kdonthi replied:
From what I understood, I just saw Transactions was a regular array of transactions, so I thought the last element from the last non-empty block would be the latest added transaction and would have the "latestPrice" (which is what Geth uses). 

Of course, this is not set in stone - would you like me to try to get the 3rd or the 4th latest transaction instead?

tkstanczak commented on src/Nethermind/Nethermind.JsonRpc/Modules/Eth/GasPrice.cs:
```diff
@@ -0,0 +1,176 @@
+        
+        private SortedSet<UInt256> eth_initializeValues(Block headBlock, Block genesisBlock, out long blocksToGoBack, out int percentile,
+            out long threshold, out long currBlockNumber, out long genBlockNumber, out UInt256? gasPriceLatest)
+        {
```
you do not need genesis block here, genesis always has number 0

tkstanczak commented on src/Nethermind/Nethermind.JsonRpc/Modules/Eth/GasPrice.cs:
```diff
@@ -0,0 +1,176 @@
+            Comparer<UInt256> comparer = Comparer<UInt256>.Create(((a, b) =>
+            {
+                int res = a.CompareTo(b);
+                return res == 0 ? 1 : res;
```
this comparer is broken because it does not follow the ordering requirement for a == b
it may lead to runtime exceptions

kdonthi replied:
The reason I wanted to use this comparer was to have a list of all the gas prices, like Geth does in suggestTipCap. A SortedList doesn't usually allow duplicates, but by setting the ordering of two equal values, I am able to allow it.

Could you help me understand when having two equal gas prices would result in a runtime error? I successfully ran a test that had equal gas prices (eth_gas_price_get_tx_from_min_blocks_if_num_tx_greater_than_or_equal_to_limit).

If still necessary, I can also create a list and sort it manually.

tkstanczak commented on src/Nethermind/Nethermind.JsonRpc/Modules/Eth/GasPrice.cs:
```diff
@@ -0,0 +1,176 @@
+        }
+        
+        private SortedSet<UInt256> eth_initializeValues(Block headBlock, Block genesisBlock, out long blocksToGoBack, out int percentile,
+            out long threshold, out long currBlockNumber, out long genBlockNumber, out UInt256? gasPriceLatest)
```
Probably better to store things like percentile as consts in the class then initialize it to a const value here?

tkstanczak commented on src/Nethermind/Nethermind.JsonRpc/Modules/Eth/GasPrice.cs:
```diff
@@ -0,0 +1,176 @@
+            threshold = blocksToGoBack * 2;
+            currBlockNumber = headBlock!.Number;
+            genBlockNumber = genesisBlock!.Number;
+            gasPriceLatest = eth_latestGasPrice(headBlock!.Number, genesisBlock!.Number);
```
you initialize blocks to go back but you do not use it inside this call, make it a consts and use it everywhere consistently

tkstanczak commented on src/Nethermind/Nethermind.JsonRpc/Modules/Eth/GasPrice.cs:
```diff
@@ -0,0 +1,176 @@
+                out int percentile, out long threshold, out long currBlockNumber, out long genBlockNumber, out UInt256? gasPriceLatest);
+            if (gasPriceLatest == null)
+            {
+                return ResultWrapper<UInt256?>.Fail("gasPriceLatest was not set properly.");
```
the error message here looks like a message for developer but it is sent out to the user

tkstanczak commented on src/Nethermind/Nethermind.JsonRpc/Modules/Eth/GasPrice.cs:
```diff
@@ -0,0 +1,176 @@
+                return ResultWrapper<UInt256?>.Fail("gasPriceLatest was not set properly.");
+            }
+            
+            gasPrices = eth_AddingPreviousBlockTx(ignoreUnder, blocksToGoBack, currBlockNumber, genBlockNumber, 
```
it is important to take into consideration not only gas prices but also gas used for the previous block and when calculating percentile, take the last 20% of block space and not the last 20% of gas prices

kdonthi replied:
Is this a change for EIP-1559 and Is this something we want to do differently than Geth?

If we do want to do it differently, should I create an array of block space (capacity - number of transactions in block) in addition to the gas used by the previous block to calculate the new gas price? 

If we call the gas used by the previous block _g_, the 20th percentile of sorted block space _s_, and the number of tx in the previous block _n_, then_ could the new gas be calculated as: (_g_ / _n_ ) * (_s_ / (capacity / 2))? 

That formula would allow us to increase the average gas price from the previous block if the number of tx is greater than half the capacity, and reduce it if it is less than half the capacity.

tkstanczak reviewed (changes requested):
Change the algorithms to take into account block space usage (gas used / gas limit)
Improve code and algorithms quality, use consts properly, think about naming of methods, improve conditions checks.

kdonthi commented:
## Changes:
- Added GasPriceOracle seam into EthRpcModule
- Refactored GasPriceOracle and EthRpcModule Tests into smaller functions to be more readable and to remove duplication
- Improved GasPriceCalculation to find 60th percentile of 3 lowest Effective Gas Price transactions from last ~20 blocks

## Types of changes

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update
- [ ] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Other (please describe): 

## Testing
**Requires testing**

- [x] Yes
- [ ] No

**In case you checked yes, did you write tests??**

- [x] Yes
- [ ] No

tkstanczak commented on src/Nethermind/Nethermind.JsonRpc/Modules/Eth/EthRpcModule.cs:
```diff
@@ -39,6 +41,7 @@
 using Nethermind.Trie;
 using Nethermind.TxPool;
 using Nethermind.Wallet;
+using Org.BouncyCastle.Security;
```
hm?

tkstanczak commented on src/Nethermind/Nethermind.JsonRpc/Modules/Eth/EthRpcModule.cs:
```diff
@@ -21,6 +21,8 @@
 using System.Text;
 using System.Threading;
 using System.Threading.Tasks;
+using Jint.Parser.Ast;
```
hm?

tkstanczak commented on src/Nethermind/Nethermind.JsonRpc/Modules/Eth/EarlyExitManager.cs:
```diff
@@ -0,0 +1,123 @@
+
+namespace Nethermind.JsonRpc.Modules.Eth
+{
+    public class EarlyExitManager
```
add a comment explaining what this class does

tkstanczak commented on src/Nethermind/Nethermind.JsonRpc/Modules/Eth/GasPriceConfig.cs:
```diff
@@ -0,0 +1,16 @@
+{
+    public static class GasPriceConfig
+    {
+        public const int NoHeadBlockChangeErrorCode = 7;
```
Add comments explaining what these settings mean For example -> Percentile can be percentile of anything

tkstanczak commented on src/Nethermind/Nethermind.JsonRpc/Modules/Eth/EarlyExitManager.cs:
```diff
@@ -0,0 +1,123 @@
+{
+    public class EarlyExitManager
+    {
+        private IBlockFinder? _blockFinder;
```
you managed to extract classes with separated responsibility and it seems a right approach but you missed an opportunity to test them in isolation -> both here and in the GasPriceOracle

tkstanczak commented on src/Nethermind/Nethermind.JsonRpc/Modules/Eth/GasPriceOracle.cs:
```diff
@@ -0,0 +1,173 @@
+            return gasPriceEstimate;
+        }
+
+        private void SetLastGasPrice(UInt256? lastGasPrice)
```
this is an unnecessary setter -> in C# you can use properties for that

tkstanczak commented on src/Nethermind/Nethermind.JsonRpc/Modules/Eth/GasPriceOracle.cs:
```diff
@@ -0,0 +1,173 @@
+        }
+        
+        private static UInt256? FindMinOfThisAndMaxPrice(UInt256? gasPriceEstimate)
+        {
```
UInt256.Min(gasPriceEstimate, GasPriceConfig._maxGasPrice)

tkstanczak commented on src/Nethermind/Nethermind.JsonRpc/Modules/Eth/GasPriceOracle.cs:
```diff
@@ -0,0 +1,173 @@
+        }
+        
+        private static UInt256 GetElementAtIndex(List<UInt256> txGasPriceList, int roundedIndex)
+        {
```
unnecessary method -> just overrides the indexer

tkstanczak replied:
use txGasPriceList[roundedIndex]

tkstanczak commented on src/Nethermind/Nethermind.JsonRpc/Modules/Eth/GasPriceOracle.cs:
```diff
@@ -0,0 +1,173 @@
+            return _blockFinder!.FindBlock(blockNumber);
+        }
+        
+        private static bool BlockExists(Block? foundBlock)
```
the method doe snot describe what it does.

tkstanczak replied:
seems not needed

tkstanczak commented on src/Nethermind/Nethermind.JsonRpc/Modules/Eth/ValidTxInsertionManager.cs:
```diff
@@ -0,0 +1,122 @@
+            Transaction[] transactionsInBlock = block.Transactions;
+            int countTxAdded;
+
+            if (TransactionsExistIn(transactionsInBlock))
```
TransactionsExistIn(block) would read better

tkstanczak commented on src/Nethermind/Nethermind.JsonRpc/Modules/Eth/ValidTxInsertionManager.cs:
```diff
@@ -0,0 +1,122 @@
+
+                if (countTxAdded == 0)
+                {
+                    AddDefaultPriceTo();
```
AddDefaultPriceTo? what does it do?

tkstanczak commented on src/Nethermind/Nethermind.JsonRpc/Modules/Eth/GasPriceOracle.cs:
```diff
@@ -0,0 +1,173 @@
+
+        public ResultWrapper<UInt256?> GasPriceEstimate(IBlockFinder blockFinder)
+        {
+            if (_blockFinder == null)
```
why do you pass this argument in the estimate method and then set it as a field?
either set it in the constructor or pass around?
otherwise this is very not thread safe?

tkstanczak commented on src/Nethermind/Nethermind.JsonRpc/Modules/Eth/GasPriceOracle.cs:
```diff
@@ -0,0 +1,173 @@
+            _blockLimit = blockLimit ?? GasPriceConfig.DefaultBlocksLimit;
+            _softTxThreshold = GasPriceConfig.SoftTxLimit;
+            _baseFee = baseFee ?? GasPriceConfig.DefaultBaseFee;
+            _addValidTx = new ValidTxInsertionManager(this, _ignoreUnder, _eip1559Enabled, _baseFee);
```
This is a field if it is mutated in the calls asking for an estimate?

tkstanczak commented on src/Nethermind/Nethermind.JsonRpc/Modules/Eth/GasPriceOracle.cs:
```diff
@@ -0,0 +1,173 @@
+                Block? block = FindBlockAtNumber(currentBlockNumber);
+                if (BlockExists(block))
+                {
+                    int txsAdded = _addValidTx.AddValidTxAndReturnCount(block!);
```
_addValidTx seems to be a strange name

tkstanczak reviewed (changes requested):
Please read inline comments.
There is some improvement in coding style but some things are very confusing.

MarekM25 commented on src/Nethermind/Nethermind.JsonRpc.Test/Modules/BlockConstructor.cs:
```diff
@@ -0,0 +1,149 @@
+                .TestObject;
+        }
+
+        public static string[][] CollectTxStrings(params string[][] txInfo)
```
Why we need this method?

MarekM25 commented on src/Nethermind/Nethermind.JsonRpc.Test/Modules/BlockConstructor.cs:
```diff
@@ -0,0 +1,149 @@
+
+        public static string[] GetTxString(string privateKeyLetter, string gasPrice, string nonce)
+        {
+            return new[] {privateKeyLetter, gasPrice, nonce};
```
And this?

MarekM25 commented on src/Nethermind/Nethermind.JsonRpc.Test/Modules/BlockConstructor.cs:
```diff
@@ -0,0 +1,149 @@
+
+        public static KeyValuePair<int, string[][]> BlockNumberAndTxStringsKeyValuePair(int blockNumber, string[][] txInfo)
+        {
+            return new KeyValuePair<int, string[][]>(blockNumber, txInfo);
```
And this? :)

MarekM25 commented on src/Nethermind/Nethermind.JsonRpc.Test/Modules/BlockConstructor.cs:
```diff
@@ -0,0 +1,149 @@
+        }
+
+        public static PrivateKey PrivateKeyForLetter(char privateKeyLetter)
+        {
```
Do not do that. We can use TestItem.PrivateKeyA. There is no sense to introduce additional method

MarekM25 commented on src/Nethermind/Nethermind.JsonRpc/Modules/Eth/HeadBlockChangeManager.cs:
```diff
@@ -0,0 +1,49 @@
+
+namespace Nethermind.JsonRpc.Modules.Eth
+{
+    //Checks edge cases where the Head block did not change from previous call of eth_gasPrice from an instance of EthRpcModule
```
It is strange because name is not suggesting that it is used for eth_gasPrice.

MarekM25 commented on src/Nethermind/Nethermind.JsonRpc/Modules/Eth/GasPriceConfig.cs:
```diff
@@ -0,0 +1,16 @@
+
+namespace Nethermind.JsonRpc.Modules.Eth
+{
+    public static class GasPriceConfig
```
I like it but maybe the better name will be EthGasPriceConstants or EthGasPriceEstimatorConstants?

but Constants not Config

MarekM25 commented on src/Nethermind/Nethermind.JsonRpc/Modules/Eth/TxInsertionManager.cs:
```diff
@@ -0,0 +1,127 @@
+
+        private static bool TransactionIsNotEip1559(Transaction transaction)
+        {
+            return !transaction.IsEip1559;
```
There is no sense in adding such a method when you can simply use !transaction.IsEip1559 somewhere. I see that you adding a lot of methods like that, so it is a general remark for this PR.

MarekM25 commented on src/Nethermind/Nethermind.JsonRpc.Test/Modules/BlockConstructor.cs:
```diff
@@ -0,0 +1,149 @@
+        }
+
+        private static IEnumerable<Transaction> ConvertToEip1559Txs(params string[][] txsInfo)
+        {
```
strange

MarekM25 commented on src/Nethermind/Nethermind.JsonRpc.Test/Modules/EthGasPriceEthRpcModuleTests.cs:
```diff
@@ -0,0 +1,314 @@
+//  Copyright (c) 2021 Demerzel Solutions Limited
```
Please rename this file to EthRpcModuleTests.GasPrice

MarekM25 commented on src/Nethermind/Nethermind.JsonRpc/Modules/Eth/GasPriceOracle.cs:
```diff
@@ -0,0 +1,148 @@
+        private readonly IHeadBlockChangeManager _headBlockChangeManager;
+
+        public GasPriceOracle(
+            bool isEip1559Enabled = false, 
```
passing this variables in constructor is super strange for me

MarekM25 reviewed (changes requested):
I don't get through all the changes. Could you read my comments and try to apply them to the whole PR?

MarekM25 commented:
Please merge it with master and ensure that all of the tests are green before adding reviewers.

MarekM25 commented on src/Nethermind/Nethermind.JsonRpc.Test/Modules/EthRpcModuleTests.GasPrice.cs:
```diff
@@ -0,0 +1,299 @@
+        }
+
+        [Test]
+        public void Eth_gasPrice_BlocksAvailableLessThanBlocksToCheck_ShouldBeSuccessful()
```
I would like to see tests like, but for eth_gasPrice here
```
            using Context ctx = await Context.Create();
            string serialized = ctx._test.TestEthRpc("eth_getBalance", TestItem.AddressA.Bytes.ToHexString(true));
            Assert.AreEqual("{\"jsonrpc\":\"2.0\",\"result\":\"0x3635c9adc5de9f09e5\",\"id\":67}", serialized);
```

MarekM25 commented on src/Nethermind/Nethermind.JsonRpc.Test/Modules/TestBlockConstructor.cs:
```diff
@@ -0,0 +1,33 @@
+using Nethermind.Core;
```
Minor: legal notes :)

MarekM25 commented on src/Nethermind/Nethermind.JsonRpc.Test/Modules/TestRpcBlockchain.cs:
```diff
@@ -123,6 +124,8 @@ protected override async Task<TestBlockchain> Build(ISpecProvider specProvider =
             ITxSealer txSealer1 = new NonceReservingTxSealer(txSigner, Timestamper, TxPool);
             TxSender ??= new TxPoolSender(TxPool, txSealer0, txSealer1);
             
+            GasPriceOracle = new GasPriceOracle(Substitute.For<ISpecProvider>());
```
Why not SpecProvider instead of Substitute?

MarekM25 commented on src/Nethermind/Nethermind.JsonRpc/Modules/Eth/EthRpcModule.cs:
```diff
@@ -58,15 +55,18 @@ public partial class EthRpcModule : IEthRpcModule
-
         private readonly ILogger _logger;
+        protected IGasPriceOracle GasPriceOracle { get; set; }
+        public Dictionary<long, Block> BlockNumberToBlockDictionary { get; private set; }
```
Can we remove this dictionary?

MarekM25 commented on src/Nethermind/Nethermind.JsonRpc/Modules/Eth/EthRpcModule.cs:
```diff
@@ -140,12 +141,39 @@ public ResultWrapper<Address> eth_coinbase()
+            }
         }
 
+        private Dictionary<long, Block> CreateBlockNumberToBlockDictionary(Block? headBlock)
```
I think we should move everything to GasPriceOracle, but you can ignore it..
But I wonder if we need to build this dict everytime? What geth is doing here?

MarekM25 commented on src/Nethermind/Nethermind.JsonRpc/Modules/Eth/GasPriceEstimateTxInsertionManager.cs:
```diff
@@ -0,0 +1,100 @@
+        private readonly UInt256 _baseFee;
+        private readonly ISpecProvider _specProvider;
+        public GasPriceEstimateTxInsertionManager(IGasPriceOracle gasPriceOracle, UInt256? ignoreUnder, 
+            UInt256 baseFee, ISpecProvider specProvider)
```
Passing baseFee sugesst doing new object for every block... you can always do it with _blockTree.Head.BaseFee

MarekM25 commented on src/Nethermind/Nethermind.JsonRpc/Modules/Eth/GasPriceEstimateHeadBlockChangeManager.cs:
```diff
@@ -0,0 +1,22 @@
+namespace Nethermind.JsonRpc.Modules.Eth
+{
+    //Checks edge cases where the Head block did not change for previous call of eth_gasPrice from an instance of EthRpcModule
+    public class GasPriceEstimateHeadBlockChangeManager : IHeadBlockChangeManager
```
Consider moving this method to another class and remove it

MarekM25 commented on src/Nethermind/Nethermind.JsonRpc/Modules/Eth/EthGasPriceConstants.cs:
```diff
@@ -0,0 +1,15 @@
+        public const int DefaultBlocksLimit = 20; //Limit for how many blocks we check txs in to add to sortedTxList
+        public const int DefaultGasPrice = 1; //Tx price added to sortedTxList for a block that has 0 txs (now adds 1 price)
+        public const int SoftTxLimit = 40; //If a block adds only 1 tx price, the block is not counted towards BlocksLimit if sortedTxList.Count() + blocksLeftToCheck <= SoftTxLimit (after price is added to sortedTxList) 
+        public const int DefaultBaseFee = 200; //Minimum price a transaction sender has to pay
```
Why DefaultBaseFee = 200?
Why _maxGasPrice = 500?

Where did you find these consts?

kdonthi replied:
Default base Fee, I found in a Geth test (I removed it now), but maxGasPrice was part of the Geth code: https://github.com/ethereum/go-ethereum/blob/master/eth/gasprice/gasprice.go#L35.

MarekM25 commented on src/Nethermind/Nethermind.JsonRpc/Modules/Eth/GasPriceOracle.cs:
```diff
@@ -0,0 +1,115 @@
+            ISpecProvider specProvider,
+            UInt256? ignoreUnder = null, 
+            int? blockLimit = null, 
+            UInt256? baseFee = null,
```
Could you explain why do you pass baseFee?

kdonthi replied:
I remembered for EIP1559, the baseFee was supposed to change according to the fullness of the block and I thought that I should start with the current baseFee and update it.

But after you asked that, I realized that this function is not responsible for calculating new baseFees; we are calculating the new EffectiveGasPrice to pay for transactions in the next block.

MarekM25 reviewed (changes requested):
Please revert it:
![image](https://user-images.githubusercontent.com/9356351/125624858-310389e5-659f-47e6-b916-5922d090c0cc.png)

kdonthi commented:
I am checking the last 20 blocks for the 3 transactions with minimum EffectiveGasPrice and adding their EffectiveGasPrices to a list (let us call this EGList).

I sort EGList in ascending order, and return the value at the 60th percentile of the maximum index.

Some notable edge cases:

1. If a block has 0 transactions, then we add the DefaultGasPrice to the EGList .
2. Remember how I said we check the last 20 blocks? We actually only count the last 20 blocks with <i>more than 1 tx </i>(so if a block only adds 1 gasPrice, it doesn’t count towards this limit; therefore we can actually check more than 20 blocks).

MarekM25 commented on src/Nethermind/Nethermind.JsonRpc/Modules/Eth/GasPriceEstimateTxInsertionManager.cs:
```diff
@@ -0,0 +1,99 @@
+        }
+
+        private bool Eip1559ModeCompatible(Transaction transaction, bool eip1559Enabled)
+        {
```
So if we have 1559Enabled we can only add 1559 tx? What about legacy tx? Could you explain it? It looks like bug

kdonthi replied:
The intention was that:
 if (Eip1559Enabled)
      1559 or regular Tx is okay,
else
      only regular tx is okay

MarekM25 replied:
Yeah. Sorry, my bad :)

MarekM25 commented on src/Nethermind/Nethermind.JsonRpc/Modules/Eth/GasPriceEstimateTxInsertionManager.cs:
```diff
@@ -0,0 +1,99 @@
+            }
+            else
+            {
+                GetTxGasPriceList().Add((UInt256) _gasPriceOracle.FallbackGasPrice!);
```
Cosmetic: I suggest doing:
var txGasPriceList = GetTxGasPriceList();
then you can use txGasPriceList instead of method.

MarekM25 commented on src/Nethermind/Nethermind.JsonRpc/Modules/Eth/GasPriceOracle.cs:
```diff
@@ -0,0 +1,124 @@
+
+        public GasPriceOracle(
+            ISpecProvider specProvider,
+            UInt256? ignoreUnder = null, 
```
Is it possible to remove ignoreUnder and blockLimit from constructor?

MarekM25 commented on src/Nethermind/Nethermind.JsonRpc/Modules/Eth/IGasPriceOracle.cs:
```diff
@@ -0,0 +1,16 @@
+{
+    public interface IGasPriceOracle
+    {
+        public ISpecProvider SpecProvider { get; }
```
Is exposing SpecProvider needed here?

MarekM25 commented on src/Nethermind/Nethermind.JsonRpc/Modules/Eth/IHeadBlockChangeManager.cs:
```diff
@@ -0,0 +1,28 @@
+using System;
+using Nethermind.Core;
+using Nethermind.Int256;
+
```
I think we can remove it now

MarekM25 commented on src/Nethermind/Nethermind.JsonRpc.Test/Modules/Eth/EthRpcModuleTests.GasPrice.cs:
```diff
@@ -0,0 +1,169 @@
+            return new[]{firstBlock, secondBlock, thirdBlock};
+        }
+        
+        public class TestEthRpcModule : EthRpcModule
```
Is the below code in this file needed?

MarekM25 commented on src/Nethermind/Nethermind.JsonRpc.Test/Modules/GasPriceEstimateTxInsertionManagerTests.TestableGasPriceEstimateTxInsertionManager.cs:
```diff
@@ -0,0 +1,48 @@
+//  along with the Nethermind. If not, see <http://www.gnu.org/licenses/>.
+// 
+
+using System.Collections.Generic;
```
I am wondering if we can achieve the same just by mocking IGasPriceEstimateTxInsertionManager

MarekM25 commented on src/Nethermind/Nethermind.JsonRpc.Test/Modules/Eth/EthRpcModuleTests.GasPrice.cs:
```diff
@@ -0,0 +1,169 @@
+        }
+
+        [Test]
+        public async Task Eth_gasPrice_BlocksAvailableLessThanBlocksToCheck_ShouldGiveCorrectResult()
```
Can you add this type of test with 1559 transactions and for all edge cases you mentioned before?

MarekM25 commented on src/Nethermind/Nethermind.JsonRpc.Test/Modules/GasPriceEstimateTxInsertionManagerTests.cs:
```diff
@@ -0,0 +1,181 @@
+using System;
```
Minor: legal notes

MarekM25 commented on src/Nethermind/Nethermind.JsonRpc.Test/Modules/GasPriceOracleTests.ShouldReturnSameGasPriceGasPriceOracle.cs:
```diff
@@ -0,0 +1,51 @@
+            {
+            }
+
+            public override bool ShouldReturnSameGasPrice(Block? lastHead, Block? currentHead, UInt256? lastGasPrice)
```
Can we mock it instaead fof creating seperate class?

MarekM25 commented on src/Nethermind/Ethereum.Blockchain.Test/SpecialTests.cs:
```diff
@@ -18,14 +18,16 @@
     {
-        [TestCaseSource(nameof(LoadTests))]
+        //[TestCaseSource(nameof(LoadTests))]
+        [Todo("Test not working in master.")]
```
We need to uncomment it. These tests are flaky, but we can't mark it with ToDo.

MarekM25 commented on src/Nethermind/Nethermind.AuRa.Test/Contract/TxPriorityContractTests.cs:
```diff
@@ -93,7 +94,7 @@ public async Task whitelist_should_return_correctly()
         }
         
-        [Test]
+        [Todo("Needs to be fixed, configuration error in master.")]
```
We need to uncomment it. These tests are flaky, but we can't mark it with to do.

MarekM25 commented on src/Nethermind/Nethermind.Core.Test/Builders/BlockTreeBuilder.cs:
```diff
@@ -213,6 +213,33 @@ public static void AddBlock(IBlockTree blockTree, Block block)
             blockTree.UpdateMainChain(new[] {block}, true);
         }
 
+        public BlockTreeBuilder WithBlocks(params Block[] blocks)
```
Looks better now!

MarekM25 commented on src/Nethermind/Nethermind.Core/InternalsVisibility.cs:
```diff
@@ -19,3 +19,4 @@
 [assembly: InternalsVisibleTo("Nethermind.Core.Test")]
 [assembly: InternalsVisibleTo("Nethermind.Blockchain.Test")]
 [assembly: InternalsVisibleTo("Nethermind.Clique.Test")]
+[assembly: InternalsVisibleTo("DynamicProxyGenAssembly2")]
```
What is that? We don't want add external libraries

MarekM25 commented on src/Nethermind/Nethermind.JsonRpc.Test/Nethermind.JsonRpc.Test.csproj:
```diff
@@ -17,6 +17,7 @@
     <PackageReference Include="NSubstitute" Version="4.2.2" />
     <PackageReference Include="NUnit" Version="3.13.2" />
     <PackageReference Include="NUnit3TestAdapter" Version="4.0.0" />
+    <PackageReference Include="RhinoMocks" Version="4.0.0-alpha3" />
```
We're using NSubstitute... Is RhinoMocks needed?

MarekM25 commented on src/Nethermind/Nethermind.JsonRpc/Modules/Eth/EthGasPriceConstants.cs:
```diff
@@ -0,0 +1,13 @@
+using Nethermind.Int256;
```
Legal notes :)

MarekM25 commented on src/Nethermind/Nethermind.JsonRpc/Modules/Eth/EthRpcModule.cs:
```diff
@@ -89,6 +87,7 @@ private static bool HasStateForBlock(IBlockchainBridge blockchainBridge, BlockHe
             _specProvider = specProvider ?? throw new ArgumentNullException(nameof(specProvider));
         }
 
+        public virtual IGasPriceOracle GasPriceOracle => new GasPriceOracle(_specProvider);
```
Please pass IGasPriceOracle through constructor :)

MarekM25 commented on src/Nethermind/Nethermind.JsonRpc/Modules/Eth/GasPriceEstimateTxInsertionManager.cs:
```diff
@@ -0,0 +1,101 @@
+using System.Collections.Generic;
```
Legal notes :)

MarekM25 commented on src/Nethermind/Nethermind.JsonRpc/Modules/Eth/GasPriceOracle.cs:
```diff
@@ -0,0 +1,131 @@
+using System;
```
Add to all files lega notes

MarekM25 commented on src/Nethermind/Nethermind.JsonRpc/Modules/Eth/IEthRpcModule.cs:
```diff
@@ -275,5 +275,11 @@ public interface IEthRpcModule : IRpcModule
             ExampleResponse = " \"accountProof\": [\"0xf90211a0446f43a2d3e433732c75bcf3519f4844e0441a4d39b31395ee9a65700c30d3b4a0b9720db63afe9909418fb6e02c9d9f225310856549cc1b66b486041f2d867250a046e6e560e52d4fe0d2f6609f489ba85f18ad1655fee18452588dc08388fbd711a01e68f36c91bd15cbf65587d6db2a7cbd6635907291e77dd80152161da9a28a48a0d2178a1891c26ccaa2d2cec82c231a0640a26a1f5e07c7b5493761bdb3aa94e5a0fa9093 ...
         ResultWrapper<AccountProof> eth_getProof([JsonRpcParameter(ExampleValue = "[\"0x7F0d15C7FAae65896648C8273B6d7E43f58Fa842\",[  \"0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421\" ],\"latest\"]")] Address accountAddress, byte[][] hashRate, BlockParameter blockParameter);
+        
+        public IGasPriceOracle GasPriceOracle
```
This interface should not have GasPriceOracle

MarekM25 reviewed (commented):
It looks better and better... :)
